Sources/X509/X509BaseTypes/TimeCalculations.swift

```diff
@@ -0,0 +1,292 @@
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+// This file contains code derived from the musl implementation at
+// https://git.musl-libc.org/cgit/musl/tree/src/time/__secs_to_tm.c and
+// https://git.musl-libc.org/cgit/musl/tree/src/time/__tm_to_secs.c.
+//
+// These implementations have been translated into Swift appropriately for this
+// use-case.
+//
+// The copyright for the original implementation is:
+//
+//----------------------------------------------------------------------
+// Copyright © 2005-2020 Rich Felker, et al.
+//
+// Permission is hereby granted, free of charge, to any person obtaining
+// a copy of this software and associated documentation files (the
+// "Software"), to deal in the Software without restriction, including
+// without limitation the rights to use, copy, modify, merge, publish,
+// distribute, sublicense, and/or sell copies of the Software, and to
+// permit persons to whom the Software is furnished to do so, subject to
+// the following conditions:
+//
+// The above copyright notice and this permission notice shall be
+// included in all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+// EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+// MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
+// IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
+// CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
+// TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
+// SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+//----------------------------------------------------------------------
+
+extension Int64 {
+    // 2000-03-01 (mod 400 year, immediately after feb29
+    @inlinable
+    static var leapoch: Int64 {
+        .secondsFromEpochToYear2000 + .secondsPerDay * (31 + 29)
+    }
+
+    @inlinable
+    static var secondsFromEpochToYear2000: Int64 {
+        946684800
+    }
+
+    @inlinable
+    static var daysPer400Years: Int64 {
+        365 * 400 + 97
+    }
+
+    @inlinable
+    static var daysPer100Years: Int64 {
+        365 * 100 + 24
+    }
+
+    @inlinable
+    static var daysPer4Years: Int64 {
+        365 * 4 + 1
+    }
+
+    @inlinable
+    static var secondsPerDay: Int64 {
+        24 * 60 * 60
+    }
+
+    @inlinable
+    static var secondsPerYear: Int64 {
+        .secondsPerDay * .daysPerYear
+    }
+
+    @inlinable
+    static var daysPerYear: Int64 {
+        365
+    }
+
+    @inlinable
+    static func daysInMonth(_ month: Int) -> Int64 {
+        // This may seem weird, but these months are indexed from _March_.
+        // Thus, month 0 is March, month 11 is February.
+        switch month {
+        case 0, 2, 4, 5, 7, 9, 10:
+            return 31
+        case 1, 3, 6, 8:
+            return 30
+        case 11:
+            return 29
```

Comment:
February usually takes `28` days. I didn't get why it was set to 29 as default.

Reply:
The use of 29 here is subtle and weird, but we get away with it because all the calculations are performed on a "year" that begins in March. We calculate the year _before_ we calculate the day of the year, and that year calculation is normalized for leap years. Thus, we can only get to the 29th day of February _if_ we already know we're in a leap year: otherwise we'll be in the first day of the following year, which starts on the 1st of March.

Reply:
Thanks for the response!